File: October/Week3/SoHyun/L11_도로보수로봇.py
def is_possible(l, x, n, k):
    cnt, last_x = 1, x[0]
    for i in range(1, n):
        if x[i] - last_x + 1 <= l:
            continue
        cnt += 1
        last_x = x[i]
    return cnt <= k

def search(s, e, x, n, k):
    if s > e:
        return x[-1] - x[0] + 1
    m = (s + e) // 2

    if is_possible(m, x, n, k):
        return min(m, search(s, m-1, x, n, k))
    else:
        return search(m+1, e, x, n, k)

n , k = map(int, input().split())
x = list(map(int, input().split()))
x.sort()

print(search(1, x[-1] - x[0] + 1, x, n, k))

File: October/Week3/Semin/L11_도로 보수 로봇.java
import java.util.*;
import java.io.*;

public class Main {
    static int N, K;
    static int[] x;

    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] NM = br.readLine().split(" ");
        N = Integer.parseInt(NM[0]);
        K = Integer.parseInt(NM[1]);
        x = new int[N];

        String[] s = br.readLine().split(" ");
        for (int i = 0; i < N; i++) {
            x[i] = Integer.parseInt(s[i]);
        }
        Arrays.sort(x);
        System.out.println(parametricSearch(1, x[N - 1] - x[0] + 1));
    }

    public static boolean isPossible(int l) {
        int cnt = 1, lastX = x[0];
        for (int i = 1; i < N; i++) {
            if (x[i] - lastX + 1 <= l) continue;
            cnt++;
            lastX = x[i];
        }
        return cnt <= K;
    }

    public static int parametricSearch(int s, int e) {
        if (s > e) return x[N - 1] - x[0] + 1;
        int m = (s + e) / 2;
        if (isPossible(m)) {
            return Math.min(m, parametricSearch(s, m - 1));
        } else {
            return parametricSearch(m + 1, e);
        }
    }
}
